In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.set_option("display.max_columns",None)
df = pd.read_csv('telco_customer_churn.csv')

In [3]:
df

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
0              No  No phone service             DSL             No   
1             Yes                No             DSL            Yes   
2             Yes                No             DSL            Yes   
3              No  No phone service             DSL            Yes   
4             Yes                No     Fiber optic             No   
...           ...               ...             ...            ...   
7038          Yes               Yes             DSL            Yes   
7039          Yes               Yes     Fiber optic             No   
7040           No  No phone service             DSL            Yes   
7041          Yes               Yes     Fiber optic             No   
7042          Yes                No     Fiber optic            Yes   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
0             Yes               No          No          No              No   
1              No              Yes          No          No              No   
2             Yes               No          No          No              No   
3              No              Yes         Yes          No              No   
4              No               No          No          No              No   
...           ...              ...         ...         ...             ...   
7038           No              Yes         Yes         Yes             Yes   
7039          Yes              Yes          No         Yes             Yes   
7040           No               No          No          No              No   
7041           No               No          No          No              No   
7042           No              Yes         Yes         Yes             Yes   

            Contract PaperlessBilling              PaymentMethod  \
0     Month-to-month              Yes           Electronic check   
1           One year               No               Mailed check   
2     Month-to-month              Yes               Mailed check   
3           One year               No  Bank transfer (automatic)   
4     Month-to-month              Yes           Electronic check   
...              ...              ...                        ...   
7038        One year              Yes               Mailed check   
7039        One year              Yes    Credit card (automatic)   
7040  Month-to-month              Yes           Electronic check   
7041  Month-to-month              Yes               Mailed check   
7042        Two year              Yes  Bank transfer (automatic)   

      MonthlyCharges TotalCharges Churn  
0              29.85        29.85    No  
1              56.95       1889.5    No  
2              53.85       108.15   Yes  
3              42.30      1840.75    No  
4              70.70       151.65   Yes  
...              ...          ...   ...  
7038           84.80       1990.5    No  
7039          103.20       7362.9    No  
7040           29.60       346.45    No  
7041           74.40        306.6   Yes  
7042          105.65       6844.5    No  

[7043 rows x 21 columns]

### Welcome to Homework. Please answer the questions below!

Score : Maximal 50.
Per Soal bobotnya 10

1. Buanglah baris dengan `tenure` samadengan 0 (jadi, data yang akan kita analisa di PR ini adalah baris-baris dengan `tenure` TIDAK samadengan 0). Ada berapa baris yang sekarang tersedia di data kita?

2. Bagi customer yang ***TIDAK MEMILIKI PARTNER*** dan ***TIDAK MEMILIKI DEPENDENTS***, berapa ***rata-rata*** dari `tenure` mereka?

3. Bagi customer yang ***MEMILIKI PARTNER*** dan ***TIDAK MEMILIKI DEPENDENTS***, berapa maksimal `MonthlyCharges` yang tercata?

4. Berapa banyak customer yang ***TIDAK MEMILIKI PARTNER*** namun ***MEMILIKI DEPENDENTS***?

5. Berapa rata-rata ***TOTAL CHARGES*** dari customer ***NON-SENIOR CITIZEN*** (Senior Citizen = 0) yang memiliki ***PHONE SERVICE*** dan memiliki ***INTERNET SERVICE FIBER OPTIC***?

#### Silakan menuliskan kode untuk menjawab 5 pertanyaan di atas...

In [4]:
df1=df.copy()

In [5]:
df1=df1[df1['tenure']!=0]

In [6]:
df1

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
0              No  No phone service             DSL             No   
1             Yes                No             DSL            Yes   
2             Yes                No             DSL            Yes   
3              No  No phone service             DSL            Yes   
4             Yes                No     Fiber optic             No   
...           ...               ...             ...            ...   
7038          Yes               Yes             DSL            Yes   
7039          Yes               Yes     Fiber optic             No   
7040           No  No phone service             DSL            Yes   
7041          Yes               Yes     Fiber optic             No   
7042          Yes                No     Fiber optic            Yes   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
0             Yes               No          No          No              No   
1              No              Yes          No          No              No   
2             Yes               No          No          No              No   
3              No              Yes         Yes          No              No   
4              No               No          No          No              No   
...           ...              ...         ...         ...             ...   
7038           No              Yes         Yes         Yes             Yes   
7039          Yes              Yes          No         Yes             Yes   
7040           No               No          No          No              No   
7041           No               No          No          No              No   
7042           No              Yes         Yes         Yes             Yes   

            Contract PaperlessBilling              PaymentMethod  \
0     Month-to-month              Yes           Electronic check   
1           One year               No               Mailed check   
2     Month-to-month              Yes               Mailed check   
3           One year               No  Bank transfer (automatic)   
4     Month-to-month              Yes           Electronic check   
...              ...              ...                        ...   
7038        One year              Yes               Mailed check   
7039        One year              Yes    Credit card (automatic)   
7040  Month-to-month              Yes           Electronic check   
7041  Month-to-month              Yes               Mailed check   
7042        Two year              Yes  Bank transfer (automatic)   

      MonthlyCharges TotalCharges Churn  
0              29.85        29.85    No  
1              56.95       1889.5    No  
2              53.85       108.15   Yes  
3              42.30      1840.75    No  
4              70.70       151.65   Yes  
...              ...          ...   ...  
7038           84.80       1990.5    No  
7039          103.20       7362.9    No  
7040           29.60       346.45    No  
7041           74.40        306.6   Yes  
7042          105.65       6844.5    No  

[7032 rows x 21 columns]

In [7]:
#mencari rata-rata tenure dari customer yang tidak punya pasangan dan tidak punya tanggungan 
df1[(df1['Partner']=="No")&(df1['Dependents']=="No")].tenure.describe()

count    3280.000000
mean       23.303354
std        21.854290
min         1.000000
25%         4.000000
50%        16.000000
75%        39.000000
max        72.000000
Name: tenure, dtype: float64

In [8]:
#mencari maksimal monthlycharges dari customer yang punya pasangan dan tidak punya tanggungan
df1[(df1['Partner']=="Yes")&(df1['Dependents']=="No")].MonthlyCharges.describe()

count    1653.000000
mean       74.977737
std        28.323444
min        18.400000
25%        56.100000
50%        81.400000
75%        97.700000
max       118.600000
Name: MonthlyCharges, dtype: float64

In [9]:
#mencari jumlah customer yang tidak memiliki pasangan dan punya tanggungan
df1[(df1['Partner']=="No")&(df1['Dependents']=="Yes")].customerID.describe()

count            359
unique           359
top       1452-KIOVK
freq               1
Name: customerID, dtype: object

In [10]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 
 17  

In [11]:
df1.TotalCharges=df1.TotalCharges.str.strip().astype('float')

In [12]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 
 17  

In [13]:
#mencari rata-rata totalcharges dari customer non senior citzen yang punya internet service Fiber optic dan phone service
df1[(df1['InternetService']=="Fiber optic")&(df1['PhoneService']=="Yes")&(df1['SeniorCitizen']==0)].TotalCharges.describe()

count    2265.000000
mean     3226.492737
std      2609.006098
min        68.500000
25%       777.300000
50%      2647.200000
75%      5552.500000
max      8684.800000
Name: TotalCharges, dtype: float64

### Silakan menjawab di bawah ini:

#### Soal 1:
Buanglah baris dengan tenure samadengan 0 (jadi, data yang akan kita analisa di PR ini adalah baris-baris dengan tenure TIDAK samadengan 0). Ada berapa baris yang sekarang tersedia di data kita?

#### Jawaban: 7032 baris

#### Soal 2:
Bagi customer yang TIDAK MEMILIKI PARTNER dan TIDAK MEMILIKI DEPENDENTS, berapa rata-rata dari tenure mereka?

#### Jawaban: rata-rata tenure nya adalah 23.303354

#### Soal 3:
Bagi customer yang MEMILIKI PARTNER dan TIDAK MEMILIKI DEPENDENTS, berapa maksimal MonthlyCharges yang tercata?

#### Jawaban: maksimal MonthlyChargesnya adalah 118.600000

#### Soal 4:
Berapa banyak customer yang TIDAK MEMILIKI PARTNER namun MEMILIKI DEPENDENTS?

#### Jawaban: jumlah customer yang tdk pnya pasangan dan punya tanggungan adalah 359 orang

#### Soal 5:
Berapa rata-rata TOTAL CHARGES dari customer NON-SENIOR CITIZEN (Senior Citizen = 0) yang memiliki PHONE SERVICE dan memiliki INTERNET SERVICE FIBER OPTIC?

#### Jawaban: rata-rata Total Charges dari customer non senior citizen yg pnya HP dan Internet adalah 3226.492737

# Deep Data Manipulation

Lakukan analisa dengan menggunakan teknik teknik data manipulation yang telah dipelajari sebanyak 2 analisa. kemudian berikan intrepetasi dan alasan anda melakukan analisa tersebut. <br>

Misalnya : <br>
Melakukan filter monthly charges diatas 500 kemudian saya akan mengalisa percentase churn berdasrkan kelompok A. Hasil temuan, ternyata kelompok tersebut memiliki percent churn sebesar %. <br>

Bobot maksimal 50 <br>
semakian baik analisa dan intrepetasi disertai alasan yang jelas akan mendapatkan score maximal

In [14]:
#Mencari perbandingan antara jumlah pelanggan yang churn dan tidak churn, mencari rata-rata berdasar monthlycharges dan tenure
#filter yang churn
df1[(df1['Churn']=="Yes")]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
2     3668-QPYBK    Male              0      No         No       2   
4     9237-HQITU  Female              0      No         No       2   
5     9305-CDSKC  Female              0      No         No       8   
8     7892-POOKP  Female              0     Yes         No      28   
13    0280-XJGEX    Male              0      No         No      49   
...          ...     ...            ...     ...        ...     ...   
7021  1699-HPSBG    Male              0      No         No      12   
7026  8775-CEBBJ  Female              0      No         No       9   
7032  6894-LFHLY    Male              1      No         No       1   
7034  0639-TSIQW  Female              0      No         No      67   
7041  8361-LTMKD    Male              1     Yes         No       4   

     PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup  \
2             Yes            No             DSL            Yes          Yes   
4             Yes            No     Fiber optic             No           No   
5             Yes           Yes     Fiber optic             No           No   
8             Yes           Yes     Fiber optic             No           No   
13            Yes           Yes     Fiber optic             No          Yes   
...           ...           ...             ...            ...          ...   
7021          Yes            No             DSL             No           No   
7026          Yes            No             DSL             No           No   
7032          Yes           Yes     Fiber optic             No           No   
7034          Yes           Yes     Fiber optic            Yes          Yes   
7041          Yes           Yes     Fiber optic             No           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
2                  No          No          No              No  Month-to-month   
4                  No          No          No              No  Month-to-month   
5                 Yes          No         Yes             Yes  Month-to-month   
8                 Yes         Yes         Yes             Yes  Month-to-month   
13                Yes          No         Yes             Yes  Month-to-month   
...               ...         ...         ...             ...             ...   
7021               No         Yes         Yes              No        One year   
7026               No          No          No              No  Month-to-month   
7032               No          No          No              No  Month-to-month   
7034              Yes          No         Yes              No  Month-to-month   
7041               No          No          No              No  Month-to-month   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
2                 Yes               Mailed check           53.85   
4                 Yes           Electronic check           70.70   
5                 Yes           Electronic check           99.65   
8                 Yes           Electronic check          104.80   
13                Yes  Bank transfer (automatic)          103.70   
...               ...                        ...             ...   
7021              Yes           Electronic check           59.80   
7026              Yes  Bank transfer (automatic)           44.20   
7032              Yes           Electronic check           75.75   
7034              Yes    Credit card (automatic)          102.95   
7041              Yes               Mailed check           74.40   

      TotalCharges Churn  
2           108.15   Yes  
4           151.65   Yes  
5           820.50   Yes  
8          3046.05   Yes  
13         5036.30   Yes  
...            ...   ...  
7021        727.80   Yes  
7026        403.35   Yes  
7032         75.75   Yes  
7034       6886.25   Yes  
7041        306.60   Yes  

[1869 rows x 21 columns]

In [15]:
#filter yang tidak churn
df1[(df1['Churn']=="No")]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
3     7795-CFOCW    Male              0      No         No      45   
6     1452-KIOVK    Male              0      No        Yes      22   
7     6713-OKOMC  Female              0      No         No      10   
...          ...     ...            ...     ...        ...     ...   
7037  2569-WGERO  Female              0      No         No      72   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
0              No  No phone service             DSL                   No   
1             Yes                No             DSL                  Yes   
3              No  No phone service             DSL                  Yes   
6             Yes               Yes     Fiber optic                   No   
7              No  No phone service             DSL                  Yes   
...           ...               ...             ...                  ...   
7037          Yes                No              No  No internet service   
7038          Yes               Yes             DSL                  Yes   
7039          Yes               Yes     Fiber optic                   No   
7040           No  No phone service             DSL                  Yes   
7042          Yes                No     Fiber optic                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
0                     Yes                   No                   No   
1                      No                  Yes                   No   
3                      No                  Yes                  Yes   
6                     Yes                   No                   No   
7                      No                   No                   No   
...                   ...                  ...                  ...   
7037  No internet service  No internet service  No internet service   
7038                   No                  Yes                  Yes   
7039                  Yes                  Yes                   No   
7040                   No                   No                   No   
7042                   No                  Yes                  Yes   

              StreamingTV      StreamingMovies        Contract  \
0                      No                   No  Month-to-month   
1                      No                   No        One year   
3                      No                   No        One year   
6                     Yes                   No  Month-to-month   
7                      No                   No  Month-to-month   
...                   ...                  ...             ...   
7037  No internet service  No internet service        Two year   
7038                  Yes                  Yes        One year   
7039                  Yes                  Yes        One year   
7040                   No                   No  Month-to-month   
7042                  Yes                  Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
0                 Yes           Electronic check           29.85   
1                  No               Mailed check           56.95   
3                  No  Bank transfer (automatic)           42.30   
6                 Yes    Credit card (automatic)           89.10   
7                  No               Mailed check           29.75   
...               ...                        ...             ...   
7037              Yes  Bank transfer (automatic)           21.15   
7038              Yes               Mailed check           84.

In [ ]:
#jumlah pelanggan yang churn 1869 orang, sedang yang tidak churn  5163 orang

In [17]:
#Mencari perbandingan rata-rata yang churn dan tidak berdasar monthlycharges
df.groupby("Churn").agg({
    "MonthlyCharges":["mean"]})

MonthlyCharges
                mean
Churn               
No         61.265124
Yes        74.441332

In [18]:
#Mencari perbandingan rata-rata yang churn dan tidak berdasar tenure
df.groupby("Churn").agg({
    "tenure":["mean"]})

tenure
            mean
Churn           
No     37.569965
Yes    17.979133

In [ ]:
Saya menganalisa jumlah pelanggan churn dan tidak,untuk mengetahui seberapa berjalannya kepuasan pelanggan terhadap perusahaan,
dimana yang tidak churn(5163 orang) lebih besar dari yang churn (1869 orang), hanya saja jumlah yang churn masih besar dan akan menjadi
analisa lebih lanjut untuk menurunkan jumlah yg churn. 
Untuk itu perlu juga diketahui rata-rata pelanggan yang churn dan tidak, berdasar monthly charges dan tenure, apakah variabel ini 
juga ada pengaruhnya dengan kondisi churn.
Dari hasil analisa berdasar monthlycharges, jumlah rata-rata yang churn (74.441332 orang) lebih besar dari pada yang tidak (61.265124 orang)
sedang yang berdasar tenure, pelanggan yang churn(17.979133 orang) lebih kecil dari yang tidak (37.569965 orang)